sources/editor/Stride.Editor/Engine/EntityExtensions.cs

```diff
@@ -20,12 +20,41 @@ public static Entity FindSubEntity(this Entity entity, Guid subEntityId)
             if (entity.Id == subEntityId)
                 return entity;
 
-            return entity.GetChildren().DepthFirst(x => x.GetChildren()).FirstOrDefault(x => x.Id == subEntityId);
+            foreach (var child in entity.Transform.Children)
+            {
+                if (child.Entity.FindSubEntity(subEntityId) is Entity e)
+                    return e;
+            }
+
+            return null;
         }
 
         public static Entity FindSubEntity(this Scene scene, Guid subEntityId)
         {
-            return scene.Entities.DepthFirst(x => x.GetChildren()).FirstOrDefault(x => x.Id == subEntityId);
+            if (scene.Entities.Count == 0)
+                return null;
+            
+            if (scene.Entities[0].EntityManager is EntityManager manager)
+            {
+                // The entity manager contains all entities of all the scenes under the root scene,
+                // it should be faster in most cases to iterate over that instead of going through the entity tree
+                var e = manager.GetEnumerator();
+                while (e.MoveNext())
+                {
+                    if (e.Current.Id == subEntityId && ReferenceEquals(scene, e.Current.Scene))
+                        return e.Current;
+                }
+            }
+            else // Slow path, go through tree recursively
+            {
+                foreach (Entity entity in scene.Entities)
+                {
+                    if (entity.FindSubEntity(subEntityId) is Entity e)
```

Comment:
Recursive call = bad. Keep using `DepthFirst` or investigate if `BreadthFirst` is faster (could be since it doesn't need to do a reverse enumeration).

Reply:
cfr. other replies

Reply:
We are talking about a single allocation for the call to `DepthFirst`. I don't see how that could be an issue.